### DISCUSSION TITLE
Implement binary search algo to find the current segment more quickly

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#138] [General]
### Discussed in https://github.com/sablierhq/v2-core/discussions/15

<div type='discussions-op-text'>

<sup>Originally posted by **paulrberg** June  3, 2022</sup>
The [current approach](https://github.com/sablierhq/v2-core/blob/80e474b3ae88e9228d1c7a43ce7410235f9f60de/src/SablierV2Pro.sol#L107-L111) to find the current segment time is to loop through the segments iteratively. When the segment array has a big length, this is gas inefficient.

We could make the implementation faster by using a [binary search](https://www.geeksforgeeks.org/binary-search/) algorithm. OpenZeppelin offers a Solidity v0.8 [implementation](https://ethereum.stackexchange.com/a/129511/24693) for it.</div>

**Comment by razgraf** [comment]
A nice hack to optimize iterations in a binary search is to have the client provide a guess. You'd still implement BS for worst case scenarios when the guess is wrong/lacking but for 99% of cases your while loop will last for 1 iteration, thus saving a ton of gas.

I proposed this same system for [Pods](https://github.com/pods-finance/contracts/blob/df74caffa02a3170d2fa9b5f32d9f17ee35882cf/contracts/amm/IVGuesser.sol#L114) and it has been working pretty well for the implied volatility calculator.

**Comment by PaulRBerg** [comment]
Brilliant. In effect, this would mean that the client (frontend app) runs the binary search algo (or the iterative one) in advance so that the user avoids paying gas for this computation.

**Comment by PaulRBerg** [comment]
Marking this as `stale` temporarily, until we validate our assumption that a binary algo would indeed be faster (see discussion in #17).
